### PR TITLE
Feature/kill unused models

### PR DESCRIPTION
**Pull request by jwilles**
# PR Type (Feature)

## Short Description

This PR adds functionality to shutdown models if they are not actively being used. A model instance timeout has been added and slurm model shutdown has been implemented. The model timeout logic has been implemented as:

1. Prior to model activation, timeout check is ignored. 
2. Once a model activates, the timeout check will commence alongside periodic model health check.
3. The timeout check will fail if the time difference between either the model instance activation and the last request served exceeds the timeout.
4. When the timeout check fails, a model shutdown request is sent to slurm via the model_service_client

### CODE REVIEW

File: web/models.py [hunk 1]
```diff
@@ -163,25 +169,30 @@ def is_healthy(self):
         if not is_healthy:
             self._model_instance.transition_to_state(ModelInstanceStates.FAILED)
         return is_healthy
+    
+    def is_timed_out(self, timeout):
+        last_event_datetime = self._model_instance.updated_at
+        last_generation = self._model_instance.last_generation()
+        if last_generation:
+            last_event_datetime = last_generation.created_at
```
Comment (xeon27) [inline web/models.py:177]: Shouldn't this be:
```suggestion
            last_event_datetime = last_generation.updated_at
```
to reflect the time when the generation was completed?

File: web/models.py [hunk 2]
```diff
@@ -299,6 +310,14 @@ def generate_activations(
     def is_healthy(self) -> bool:
         return self._state.is_healthy()
 
+    def is_timed_out(self, timeout):
+        return self._state.is_timed_out(timeout)
+    
+    def last_generation(self):
+        last_generation_query = db.select(ModelInstanceGeneration).where(ModelInstanceGeneration.model_instance_id == self.id).order_by(ModelInstanceGeneration.created_at.desc())
```
Comment (xeon27) [inline web/models.py:317]: If the above comment holds, this should also change to the following right?
```suggestion
        last_generation_query = db.select(ModelInstanceGeneration).where(ModelInstanceGeneration.model_instance_id == self.id).order_by(ModelInstanceGeneration.updated_at.desc())
```

**Review by xeon27 — CHANGES_REQUESTED** [review]

File: web/models.py (hunk 1 above)
Comment (markcoatsworth) [reply web/models.py:177]: I'm pretty sure these values will be the same and so it doesn't matter?

File: web/models.py [hunk 3]
```diff
@@ -110,9 +115,10 @@ def is_healthy(self):
         if not is_healthy:
             self._model_instance.transition_to_state(ModelInstanceStates.FAILED)
         return is_healthy
+    
+    def is_timed_out(self, timeout):
+        return False
```
Comment (markcoatsworth) [inline web/models.py:120]: Do we not want to check for timeouts while the model is loading? If something gets stuck while loading (say we landed on a bad node) don't we eventually want to shut it down?

**Review by markcoatsworth — APPROVED** [review]
Couple minor comments but otherwise I think this looks good.

File: web/models.py (hunk 3 above)
Comment (jwilles) [reply web/models.py:120]: It's a great point but I think that is a different problem than what the timeout is solving - which is shutting down unused models. That functionality should likely be added to the health check instead.

File: web/models.py (hunk 1 above)
Comment (xeon27) [reply web/models.py:177]: Ohh, I thought `created_at` is creation time of generation request object and `updated_at` is when the `generation` attribute is assigned, making them different based on generation time.

**Pull request merged**